File: Mock_Mar24/c.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Edge {
    int u, v, dist;
    bool operator<(const Edge& other) const {
        return dist < other.dist;
    }
};

int find(vector<int>& parents, int node) {
    if (parents[node] == node) return node;
    return parents[node] = find(parents, parents[node]);
}

void union_uv(vector<int>& parents, int u, int v) {
    parents[find(parents, u)] = find(parents, v);
}

int main(){
    int n;
    cin >> n;

    vector<Edge> edges;
    vector<int> parents(n + 1);
    for (int i = 1; i <= n; i++) {
        parents[i] = i;
    }

    for (int u = 1; u <= n; u++) {
        for (int v = 1; v <= n; v++) {
            int dist;
            cin >> dist;
            if (u != v) {
                edges.push_back({u, v, dist});
            }
        }
    }

    sort(edges.begin(), edges.end());

    for (const Edge& edge : edges) {
        if (find(parents, edge.u) != find(parents, edge.v)) {
            union_uv(parents, edge.u, edge.v);
            cout << edge.u << " " << edge.v << '\n';
        }
    }
}

File: Mock_Mar24/c.py
import sys
# from operator import itemgetter
import heapq

def find(node):
    if parents[node] != node:
        parents[node] = find(parents[node])
    return parents[node]
def union(node1, node2):
    parent1 = find(node1)
    parent2 = find(node2)
    if parent1 != parent2:
        if rank[parent1] > rank[parent2]:
            parents[parent2] = parent1
        elif rank[parent1] < rank[parent2]:
            parents[parent1] = parent2
        else:
            parents[parent2] = parent1
            rank[parent1] += 1


input = sys.stdin.read
data = input().split()
index = 0

n = int(data[index])
index += 1

heap = []
for u in range(1, n + 1):
    for v in range(1, n + 1):
        dist = int(data[index])
        index += 1
        if u < v:
            heapq.heappush(heap, (dist, u, v))

parents = list(range(n+1))
rank = [0] * (n + 1)

while heap:
    dist, u, v = heapq.heappop(heap)
    if find(u) != find(v):
        union(u, v)
        print(u, v)